packages/firestore/src/core/query.ts

```diff
@@ -47,6 +52,7 @@ export class Query {
     readonly explicitOrderBy: OrderBy[] = [],
     readonly filters: Filter[] = [],
     readonly limit: number | null = null,
+    readonly limitType: LimitType = LimitType.First,
```

Comment:
Talking to @wilhuff I think we want to start trying to differentiate `Query` and `Target`, where `Query` represents any query the SDK supports (including limit-to-last queries and in theory could include local "and", "or", etc. queries in the future) and `Target` represents only queries that the backend supports (and corresponds directly to the `Target` proto in the wire protocol).

To fully realize that, we could create a new `Target` type and replace many (most?) usages of `Query` with that new type, but in the short-term I think we could just have a `Query.toTargetQuery()` method and add a `isTargetQuery` flag.  When you call `.toTargetQuery()` it'll translate the query as necessary (in particular the limit-to-last translation) and set the `isTargetQuery` flag to true.

We can then add assertions in the rest of our code to verify `isTargetQuery` is true in places where we expect to be dealing with backend queries, in particular we should assert in:
1. The QueryData constructor  (and we should eventually rename QueryData to TargetData)
2. The IndexedDb and Memory implementations of `QueryCache.getQueryData()`.  This will ensure that we are only storing (and generating canonicalIds for) the Target version of the query.  FWIW- If we wanted we could rename QueryCache to TargetCache in the future.
3. `Serializer.ToQueryTarget()`.  This ensures we only send the Target version of the query to the backend or write it to persistence.

And once you add those asserts you'll need to presumably fix the calling code to properly do the conversions so that the asserts pass...  and there may be other places we can/should assert.

I haven't fully explored this so if this ends up messy / complicated / whatever, feel free to holler.

Reply:
One place `Query` should be used is what `View` references to. Right now it is constructed with `QueryData.query` and it will become a target query once we introduced such notion. But it needs to be still a `Query`, such that we keep the wanted order, is that right?

Reply:
That sounds right.  We will need to somehow initialize it with the original query instead of queryData.query, I guess...  but if that ends up too hard / awkward, holler.